src/rust/engine/src/watch.rs

```diff
@@ -0,0 +1,190 @@
+            if let notify::ErrorKind::PathNotFound = err.kind {
+              warn!("Path(s) did not exist: {:?}", err.paths);
+              continue;
+            } else {
+              error!("File watcher failing with: {}", err);
+              break;
+            }
+          }
+          Err(RecvTimeoutError::Timeout) => continue,
+          Err(RecvTimeoutError::Disconnected) => {
+            // The Watcher is gone: we're done.
+            break;
+          }
+        };
+      }
+      warn!("Watch thread exiting.");
+      // Signal that we're exiting (which we would also do by just dropping the channel).
+      let _ = thread_liveness_sender.send(());
+    });
+
+    Ok(InvalidationWatcher {
+      watcher,
+      liveness: thread_liveness_receiver,
+    })
+  }
+
+  ///
+  /// Watch the given path non-recursively.
+  ///
+  pub fn watch(&self, path: PathBuf) -> BoxFuture<(), ()> {
+    WatchFuture::new(self.watcher.clone(), path.clone())
+      .map_err(move |e| warn!("Failed to watch path {:?}, with error {:?}", path, e))
+      .to_boxed()
+  }
+
+  ///
+  /// Returns true if this InvalidationWatcher is still valid: if it is not valid, it will have
+  /// already logged some sort of error, and will never restart on its own.
+  ///
+  pub fn running(&self) -> bool {
+    match self.liveness.try_recv() {
+      Ok(()) | Err(TryRecvError::Disconnected) => false,
+      Err(TryRecvError::Empty) => true,
+    }
+  }
+
+  pub fn invalidate(graph: &Graph<NodeKey>, paths: &HashSet<PathBuf>, caller: &str) -> usize {
+    let InvalidationResult { cleared, dirtied } = graph.invalidate_from_roots(move |node| {
+      if let Some(fs_subject) = node.fs_subject() {
+        paths.contains(fs_subject)
+      } else {
+        false
+      }
+    });
+    // TODO: The rust log level is not currently set correctly in a pantsd context. To ensure that
+    // we see this even at `info` level, we set it to warn. #6004 should address this by making
+    // rust logging re-configuration an explicit step in `src/python/pants/init/logging.py`.
+    warn!(
+      "{} invalidation: cleared {} and dirtied {} nodes for: {:?}",
+      caller, cleared, dirtied, paths
+    );
+    cleared + dirtied
+  }
+}
+
+struct WatchFuture {
+  watcher: Arc<Mutex<RecommendedWatcher>>,
+  path: PathBuf,
+  watch_result: Arc<Mutex<Option<NotifyResult<()>>>>,
+  thread_started: bool,
+}
+
+impl WatchFuture {
+  fn new(watcher: Arc<Mutex<RecommendedWatcher>>, path: PathBuf) -> WatchFuture {
+    WatchFuture {
+      watcher,
+      path,
+      watch_result: Arc::new(Mutex::new(None)),
+      thread_started: false,
+    }
+  }
+}
+
+impl Future for WatchFuture {
+  type Item = ();
+  type Error = String;
+
+  fn poll(&mut self) -> futures::Poll<Self::Item, Self::Error> {
+    if !self.thread_started {
+      thread::spawn({
+        let current_task = futures::task::current();
+        let watcher = self.watcher.clone();
+        let watch_result2 = self.watch_result.clone();
```

Comment:
Why the trailing `2`?

Reply:
old variable before I moved stuff into local scopes. Will fix.